Remove global issues page and start using popup instead

balsa-asanovic opened the pull request:
## Problem

"We have agreed to use the sidebar for actions related to the installer itself (e.g., diagnosis tools, display lang, etc). Everything related to the target system to install should not be placed in the siderbar.

The only thing related to the target system still present in the sidebar is the issues link. The issues should be globally accessible from another place (footer, header, ...)."

As described in #872 

This PR fixes #872 

## Solution

Issues page link is removed from the sidebar and all other places where it was present.

During the progress of this PR, decision was made to drop the showing of global (all) issues at once in the popup.

Because of this, the Issues popup (previously a page) can no longer be accessed from the sidebar (or from the icon in the header, which was an idea at some point).

The Issues Dialog (popup) now shows issues related only to a single section (software, product, storage, ...) and is opened by clicking on the warning link in the sections on the overview page (and a few other places).

For this reason, the prop id for the Section component was introduced, so the issues could be filtered to a specific type for the popup based on from where has the user clicked the warning.

## Testing

-  Tested manually
- Adjusted unit test files to some extent, possibly more work is needed here.


## Screenshots

![294189450-e2b43bbc-65cb-42bb-aa64-092eb8e2b0f5](https://github.com/openSUSE/agama/assets/112288843/0177753e-c96b-42d4-a10e-7543b570104a)

dgdavid commented:
@balsa-asanovic,

Just a comment to say **thank you**. I'll do a first review ASAP. Not sure if tomorrow I'll have time, but for sure it's in the top of my todo list for the beginning of next week.

balsa-asanovic commented on web/src/components/core/Sidebar.jsx:
```diff
@@ -129,21 +130,30 @@ export default function Sidebar ({ children }) {
+              aria-label={_("Show issues popup")}
+              variant="warning"
+              icon={<Icon name="warning" />}
+              onClick={() => setShowWarningPopup(prev => !prev)}
```
`onClick={() => setShowWarningPopup(true)}` is probably better, since this is always used to open the popup and can't be used to close it.

dgdavid commented on web/src/components/core/Section.jsx:
```diff
@@ -97,7 +100,7 @@ export default function Section({
       <div className="stack content">
         {errors?.length > 0 &&
-          <ValidationErrors errors={errors} title={`${title} errors`} />}
+          <ValidationErrors errors={errors} sectionName={title} />}
```
What about using `sectionId` instead? BTW, this made me realize that maybe sections should have an `id` too, but that's could be something for #892

dgdavid reviewed (commented):
@balsa-asanovic 

First of all, grab a cup of tea or coffee... just kidding :stuck_out_tongue_winking_eye: 

Once again, my apologies for the delay in reviewing your PR. The review was blocked because I was interested in reviewing it, but I've been busy. 

Please, do not run away. Some of the comments are rather long, but that's just a reflection of how much I'm able to speak and how bad I'm at summarizing :sweat_smile: :sweat_smile: In addition, there are a few comments not directly related to your changes. You simply picked a task that ending up revealing few things we have to improve plus, as we realized yesterday in a team meeting, we lack of guidelines on how to do these things we already know/have decided how to do. E.g., when to use a skeleton and when not.

I hope you will not feel overwhelmed :pray:  Feel free to comment back if you have another perspective or think I missed something.

dgdavid commented:
BTW, I forgot to do a mention about the sections header sizes in the dialog. They looks bigger than the dialog tittle, wich... looks weird. Not your fault, of course :)

But we can try to make them looks a bit better by adding few **temporary** styles like,

<details>
<summary>Click to show/hide a diff</summary>

---

```diff
diff --git a/web/src/assets/styles/blocks.scss b/web/src/assets/styles/blocks.scss
index 0498607f..115ae636 100644
--- a/web/src/assets/styles/blocks.scss
+++ b/web/src/assets/styles/blocks.scss
@@ -339,3 +339,16 @@ span.notification-mark[data-variant="sidebar"] {
     text-align: end;
   }
 }
+
+[role="dialog"] {
+  section:not([class^="pf-c"]) {
+    > svg:first-child {
+      block-size: 24px;
+      inline-size: 24px;
+    }
+
+    h2 {
+      font-size: var(--fs-h3);
+    }
+  }
+}
diff --git a/web/src/assets/styles/variables.scss b/web/src/assets/styles/variables.scss
index 670e8751..61fe3031 100644
--- a/web/src/assets/styles/variables.scss
+++ b/web/src/assets/styles/variables.scss
@@ -11,6 +11,7 @@
   --fs-base: 14px;
   --fs-h1: 1.5rem;
   --fs-h2: 1.2rem;
+  --fs-h3: 1rem;
 
   --lh-normal: 1.5;
   --lh-medium: 1.6;
diff --git a/web/src/components/core/IssuesPage.jsx b/web/src/components/core/IssuesPage.jsx
index 2d7ee678..31b7f477 100644
--- a/web/src/components/core/IssuesPage.jsx
+++ b/web/src/components/core/IssuesPage.jsx
@@ -227,7 +227,7 @@ export default function IssuesPage({ close, sectionHighlight = "" }) {
   }, [client, load, update]);
 
   return (
-    <Popup isOpen title={_("Issues")}>
+    <Popup isOpen title={_("Issues")} data-content="issues-summary">
       <If
         condition={isLoading}
         then={<SectionSkeleton numRows={4} />}
```
</details>


|Before | After |
|-|-|
| ![Screenshot from 2023-11-29 09-29-08](https://github.com/openSUSE/agama/assets/1691872/e981942c-93a9-4323-9ea5-440115497b19) | ![Screenshot from 2023-11-29 09-28-54](https://github.com/openSUSE/agama/assets/1691872/6457ae0d-3907-48bc-980f-cd5a258e408b) |

Or something better, of course.

I wrote **temporary** because these styles should be improved, but it's out of the scope of this PR. I realized that we should use `--font-size-xs`, `--font-size-s`, etc instead of `--fs-hX`. The same for icons sizes. But we can follow the current naming  and fix it later. In fact, I plan to work in a followup of #892 for improving this kind of things for icons.

balsa-asanovic commented:
> Please, do not run away.

Oh, no way, I'll be addressing your comments ASAP.

Thanks a lot for the review, as always.

Also, I do hope that these PRs and reviews don't actually end up being more work for you, than just making it yourself. 😓

balsa-asanovic commented:
Just a quick question.

After reading through the review comments I ended up being unsure if for the Section component we should go with HTML id attribute or use ref hook (by forwarding it as a prop) which is a more React way?

I might prefer slightly using forward ref, because it evades use of id HTML attribute, but I'd gladly implement either one.

dgdavid commented:
> 
> Also, I do hope that these PRs and reviews don't actually end up being more work for you, than just making it yourself. 😓

That is by no means the case! We are happy to receive contributions! They have a lot of advantages, such as moving blocked things on (believe me, this task wouldn't have been completed in such a short period of time if you hadn't jumped on it), learning from other points of view, training us as mentors :stuck_out_tongue_winking_eye:, discovering weaknesses in the code... in addition to getting in touch with the community and meeting people who are interested in the project and/or its stack. Hence, do not worry, we do not think we're making a special effort to **help people help us**. Moreover, (open)SUSE is committed to open source philosophy.

On top of that, I'm sure I can speak for my team when I say the above.

dgdavid commented:
> Just a quick question.
> 
> After reading through the review comments I ended up being unsure if for the Section component we should go with HTML id attribute or use ref hook (by forwarding it as a prop) which is a more React way?
> 
> I might prefer slightly using forward ref, because it evades use of id HTML attribute, but I'd gladly implement either one.

Actually I was thinking in both, the `id` for telling the dialog from where it was open and the `ref` for letting React work with the section. In other words, I'm proposing to use an `id` instead of the section name/title when opening the issues dialog from a section.

If you think the `id` is not needed or it's simpler to go ahead with the name/title for now (e.g., to avoid a conflict with #892), please just go ahead as if I hadn't mentioned the `id`.

dgdavid commented:
Hey @balsa-asanovic, 

How is going? Do you need help from our side?

I'm not in a hurry, of course. Just asking in case you feel blocked or so. Please, do not forget that we're here to answer any question you might have. Don't be shy :wink: 

BTW, I'm working in a rather big PR (TL;DR: going for another strategy to work with the layout and, hopefully, stop using the react-teleporter dependency; nothing wrong with it, just changing our mind as the project evolves). Check #925 to see the progress. Letting you know about it because most probably my changes there will cause a lot of conflicts with this PR and others like https://github.com/openSUSE/agama/pull/917. But please, don't be scared. I'm going to take care of them once I finish the work.

Thanks!

balsa-asanovic commented:
Hey @dgdavid, 

> How is going? Do you need help from our side?

All is fine, I don't need any help (at least for now 😄 ).

I've already done a few things, just didn't get to finish, my schedule been a bit out of whack 😅 

> Please, do not forget that we're here to answer any question you might have. Don't be shy 😉

Of course, I will

> But please, don't be scared. I'm going to take care of them once I finish the work.

No worries, fire away.
After this PR is done, I could help you with that one, if the help is need at that point 😄

dgdavid commented:
@balsa-asanovic,

#925 has been merged, please have a look when you've time.

Meanwhile, do you mind if I Merge `master` and solve conflicts here?

balsa-asanovic commented:
Hey @dgdavid,

Sure, go ahead.

Sorry for the delay again, I've been on vacation for the last week and I just couldn't bring myself to do some work 😅

dgdavid commented:
> Sorry for the delay again, I've been on vacation for the last week and I just couldn't bring myself to do some work 😅

No problem at all. I hope you have enjoyed them.

dgdavid commented:
> Hey @dgdavid,
> 
> Sure, go ahead.

Done, although I couldn't sync the `master` branch of your fork. Only `extract-issues` one.

I'm going to add some comments.

dgdavid commented on web/src/components/core/Page.jsx:
```diff
@@ -218,6 +248,7 @@ const Page = ({ icon, title = "Agama", children }) => {
       </footer>
 
       <Sidebar isOpen={sidebarOpen} onClose={closeSidebar} />
+      <If condition={issuesDialogOpen} then={<IssuesDialog onClose={closeIssuesDialog} />} />
```
Maybe we can go for another strategy here: to keep the `IssuesDialog` always mounted but receiving a `isOpen` prop to know if it should be rendered or not. Similar to Popup/Modal. 

Not sure, honestly, but somehow it looks clean and better to me at first thought.


```suggestion
      <IssuesDialog isOpen={issuesDialogOpen}  onClose={closeIssuesDialog} />
```

dgdavid commented on web/src/components/core/Page.jsx:
```diff
@@ -94,6 +95,30 @@ const BackAction = () => {
+ * Internal component for displaying a button to show IssuesDialog when there
+ * are issues
+ *
+ * FIXME: move to its own component and import it instead (?)
```
But maybe it should follow a similar approach as Popup and Page. I.e., to move it to `IssuesDialog` component and have a `<IssuesDialog.OpenButton />` or so.

@imobachgs, what do you think about it? Just looking for another view before going ahead.

dgdavid commented on web/src/components/core/Sidebar.jsx:
```diff
@@ -129,21 +130,30 @@ export default function Sidebar ({ children }) {
   return (
     <>
       <AppActions>
```
This is no longer valid after merging #925. Now it's in core/Page component.

balsa-asanovic commented:
Hey @dgdavid,

I am finally back home 😄 

I think I've addressed all the stuff mentioned through the comments, check it out. It's possible I skipped something.

Also, regarding the id question for the sections, I left this as is, since it wasn't clear to me what is the best choice and the forward ref is now used.

dgdavid commented:
Thanks a lot!

Will review it during the day. But please, do not forget to remove the `Show issues` link from sidebar. Maybe I re-introduced it when rebasing #925 last week.

balsa-asanovic commented:
> But please, do not forget to remove the Show issues link from sidebar. Maybe I re-introduced it when rebasing https://github.com/openSUSE/agama/pull/925 last week.

Oh, you are right, It was back.

I just removed it.

balsa-asanovic commented:
```
  const IssuesWarning = () => {
    const IssuesLink = ({ text }) => {
      return (
        <Button variant="link" isInline onClick={() => navigate("/issues")}>
          {text}
        </Button>
      );
    };

    // TRANSLATORS: the installer reports some errors,
    // the text in square brackets [] is a clickable link
    const [msgStart, msgLink, msgEnd] = _("There are some reported issues. \
Please, check [the list of issues] \
before proceeding with the installation.").split(/[[\]]/);

    return (
      <p className="bold">
        {msgStart}
        <IssuesLink text={msgLink} />
        {msgEnd}
      </p>
    );
  };
```

I guess this part in InstallButton.jsx should also be updated to use the popup instead of navigating to a page which no longer exists?

dgdavid commented:
@balsa-asanovic,

I know there's a good chance you'll want to kill us... but after realizing that maybe we were wrong about the "Show issues" button placement, @imobachgs and I have invested some time to think about it and below are our new conclusions:

* The "Issues" term is wrong. We use that word just because it's one of the terms used internally to distinguish some _problems_ from errors. So, let's call them just "problems" by now.
* At this time, actually does not make sense either, to show that icon/button in the header nor to show all the problems found. Instead, we think now that Agama should

  - Stop using the "Show issues" global action
  - In the dialog, show only problems found for the section the user clicked on. I.e., if the user clicks on the "3 problems found" of "Storage" section, the popup dialog must show only these 3 problems
  - Stop using the "warning" icon in the "N problems found" links. It adds none value.
  - Remove the link used in the confirmation dialog for navigating to the issues page, which does not exist anymore.
  - Reword the text used in the confirmation dialog for "warning" about some problems present and encourage users to check them before proceeding with the installation.
  - [Out of the scope of this PR] Set the primary action in the confirmation dialog accordingly. I.e., we should set "Cancel" as the primary action if we're encouraging the user to do something else before going ahead. Of course, this is completely out of the scope of this PR.

We think that all of the above will make everything simpler: the code and the user experience. Also, we fully understand if you feel befuddled. Please, do not hesitate asking us collaboration to make these new changes before giving up. In any case, I'd like also to know your thoughts about these changes, just in case we're overlooking something important enough for keeping the "Issues" page alive.


---

/cc @imobachgs, please just check if I forgot to mention something.

dgdavid commented:
> I guess this part in InstallButton.jsx should also be updated to use the popup instead of navigating to a page which no longer exists?

You were faster than me, please check my [previous comment](https://github.com/openSUSE/agama/pull/886#issuecomment-1869546574)

dgdavid commented:
> > I guess this part in InstallButton.jsx should also be updated to use the popup instead of navigating to a page which no longer exists?
> 
> You were faster than me, please check my [previous comment](https://github.com/openSUSE/agama/pull/886#issuecomment-1869546574)

Also, note that I merged some changes this morning (#960)

dgdavid commented:
> We think that all of the above will make everything simpler: the code and the user experience.

No highlights, no refs, no checks to decide which section includes, etc.

balsa-asanovic commented:
> I know there's a good chance you'll want to kill us...

Haha, don't worry, I still not dispatching my assassins 😄 

> We think that all of the above will make everything simpler: the code and the user experience.

After a quick read through, I fully agree, it will make both the code and experience simpler.

I'll try already today to address these changes and I'll let you know if I need any help.

balsa-asanovic commented:
> The "Issues" term is wrong. We use that word just because it's one of the terms used internally to distinguish some problems from errors. So, let's call them just "problems" by now.

This got me thinking that the use of word Issues in Issues dialog, like for the title, for the arias and for the component name itself should possibly be changed?

I wasn't sure about this, so I didn't update it, I'd need further advice there.

balsa-asanovic commented:
> @balsa-asanovic,
> 
> I know there's a good chance you'll want to kill us... but after realizing that maybe we were wrong about the "Show issues" button placement, @imobachgs and I have invested some time to think about it and below are our new conclusions:
> 
>     * The "Issues" term is wrong. We use that word just because it's one of the terms used internally to distinguish some _problems_ from errors. So, let's call them just "problems" by now.
> 
>     * At this time, actually does not make sense either, to show that icon/button in the header nor to show all the problems found. Instead, we think now that Agama should
>       
>       * Stop using the "Show issues" global action
>       * In the dialog, show only problems found for the section the user clicked on. I.e., if the user clicks on the "3 problems found" of "Storage" section, the popup dialog must show only these 3 problems
>       * Stop using the "warning" icon in the "N problems found" links. It adds none value.
>       * Remove the link used in the confirmation dialog for navigating to the issues page, which does not exist anymore.
>       * Reword the text used in the confirmation dialog for "warning" about some problems present and encourage users to check them before proceeding with the installation.
>       * [Out of the scope of this PR] Set the primary action in the confirmation dialog accordingly. I.e., we should set "Cancel" as the primary action if we're encouraging the user to do something else before going ahead. Of course, this is completely out of the scope of this PR.
> 
> 
> We think that all of the above will make everything simpler: the code and the user experience. Also, we fully understand if you feel befuddled. Please, do not hesitate asking us collaboration to make these new changes before giving up. In any case, I'd like also to know your thoughts about these changes, just in case we're overlooking something important enough for keeping the "Issues" page alive.
> 
> /cc @imobachgs, please just check I didn't forget to mention something.

Covered this, @dgdavid  check it out when you find the time.

dgdavid commented on web/src/components/core/InstallButton.jsx:
```diff
@@ -23,25 +23,19 @@ import React, { useState } from "react";
-          {msgStart}
-          <Link to="/issues">{msgLink}</Link>
-          {msgEnd}
+          { _("There are some reported issues. \
+          Please review any identified concerns in the previous steps before proceeding with the installation.") }
```
@imobachgs, If you don't mind, please have a look to the text. They are not my strong skill :P 

Would be nice to have @mvidner and/or @shundhammer, but I wouldn't wait until they are back. We can refine these texts later, if needed.

imobachgs replied:
What about replacing "identified concerns" with just "them"? Anyway, we know what we are talking about.

shundhammer replied:
Not sure if a user who doesn't speak English very well will understand that. Don't forget that there are _many_ of them. I'd vote to greatly simplify that statement.

It starts with "issues" which has become en vogue in recent years (it wasn't always like that) as an euphemism for "problems". If this is about problems (or potential problems), let's just say it. If not, can we make it clearer what kind of "issues" are meant here?

As for the rest, I am not sure if it's clear to users (even native speakers) what that means, what to do, and what to look for.  I'd prefer something much clearer like "please go back and check the problems outlined in red and make the necessary changes".

We very often (starting with YaST2 in the early 2000s) had the problem of very vague, very generic messages. That typically comes from generalizing some dozen specific messages into one very generic one; generic to the point of being very much useless and just confusing the user.

balsa-asanovic replied:
A really nice point, I suppose a more straightforward message is always a better choice.

> "please go back and check the problems outlined in red and make the necessary changes"

Something like this seems fine, but I didn't update it now, waiting for others to jump in.

dgdavid replied:
As discussed today in a team daily call, I got what @shundhammer said (thanks a lot for jumping in during your vacation days) but I'm not fully agree on linking the message with the look & feel. Today is red, tomorrow who knows. But more important, there are people that are not able to distinguish colors, reason why color add value only for a group of people.

In any case, we've somehow agreed on looking for a better term and explanation, but in a different issue. So, @balsa-asanovic please keep it as it is for now in order to not make this PR longer than it is already. There are some work to do here too, but we're close to merge it. Leave the message for another PR, once we've it clear.

dgdavid commented on web/src/components/core/InstallButton.jsx:
```diff
@@ -63,11 +57,14 @@ according to the provided installation settings.") }
       </div>
       <Popup.Actions>
-        <Popup.Confirm onClick={onAccept}>
+        <Popup.PrimaryAction key="cancel" onClick={onClose} autoFocus>
+          {/* TRANSLATORS: button label */}
+          {_("Cancel")}
+        </Popup.PrimaryAction>
+        <Popup.SecondaryAction key="confirm" onClick={onAccept}>
```
I think I didn't explain well this. `Cancel` should be the primary, autofocused action only if there are issues. So, a bit more complicated than this.

```jsx

const Cancel = hasIssues ? Popup.PrimaryAction : Popup.SecondaryAction;
const Continue = hasissues ? Popup.SecondaryAction : Popup.PrimaryAction;

...

<Cancel key="cancel" onClick={onClose} autoFocus={hasIssues}>...</Cancel>
<Confirm key="confirm" onClick={onAccept} autoFocus={!hasIssues}>...</Confirm>

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -78,39 +77,53 @@ const IssueItems = ({ issues = [] }) => {
+ * @param {string} [props.openFrom] - A string which indicites which issues section should be highlighted.
  */
-const IssuesSections = ({ issues }) => {
+const IssuesSections = ({ issues, openFrom = "" }) => {
```
In general, I think we can go more _aggressive_ here and drop all the logic for building sections conditionally. I mean, the `IssuesDialog` should receive the `key`, `source`, or similar prop and retrieve the issues only for such "key", "source", "section", whatever. Thus, the IssuesDialog can be internally simplified a lot, just as we are doing visually.


```jsx

const issues = issues[key] || [];

...

return  issues.length ? <IssuesList /> : <NoIssuesFound>

```
No _stack_ needed, no `<If>`s needed, no `<Section>`'s `title` needed (but please, still using a `<Section>` for wrapping the content).

Let me know if you get what I mean.

---

// cc @imobachgs maybe the `IssuesDialog` title should be given via prop and should be something like "Software issues", "Storage issues" or "Software problems", "Storage problems", or so. Again, we need your feedback here.

imobachgs replied:
Using "Software issues", "Storage issues", and so on looks good enough, although I think we should discuss whether "issue" is a good term. But we can postpone that.

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -138,21 +152,32 @@ const IssuesContent = ({ issues }) => {
 
 /**
  * Page to show all issues.
+ *
+ * It initially shows a loading state,
+ * then fetches and displays a list of issues grouped by categories such as 'product', 'storage', and 'software'.
+ *
+ * It uses a Popup component to display the issues, and an If component to toggle between
+ * a loading state and the content state.
+ *
  * @component
+ *
+ * @param {object} props
+ * @param {boolean} [props.isOpen] - A boolean value used to determine wether to show the popup or not.
+ * @param {function} props.onClose - A function to call when the close action is triggered.
+ * @param {string} [props.openFrom] - A string which indicites which issues section should be highlighted.
```
Documentation a bit outdated (e.g., there is not highlight anymore)

dgdavid commented on web/src/components/core/Section.jsx:
```diff
@@ -52,22 +52,26 @@ import { ValidationErrors } from "~/components/core";
+ * @type {React.FC<SectionProps>}
  */
-export default function Section({
+const Section = forwardRef(function({
```
I have doubts if we still needing the forwardRef in the Section component. Probably not by now. If so, do you mind dropping it until really needed?

dgdavid reviewed (commented):
A few comments more. Please check them.

balsa-asanovic commented:
Hey, I've done some sort of simplification of the IssuesDialog component.

However, I am not sure if it looks alright now 😓 

![image](https://github.com/openSUSE/agama/assets/112288843/b7625349-5f70-41c2-867d-67aea0f8afb4)

Also, unsure about some parts I've done, for example:

`const titles = {
    software: {
      text: _("Software issues"),
      icon: () => <Icon name="apps" />
    },
    product: {
      text: _("Product issues"),
      icon: () => <Icon name="inventory_2" />
    },
    storage: {
      text: _("Storage issues"),
      icon: () => <Icon name="hard_drive" />
    }
  };`

Is this the right way to chose the title? Is there really a need for an icon?

`const IssuesSections = ({ issues }) => {
  return (
    <Section aria-label="issues-section">
      <IssueItems issues={issues} />
    </Section>
  );
};`

I've tried to simplify this part as much as possible and therefore I've removed the other prop sectionId (previously openFrom).
But, I am not sure if it is maybe needed for the appropriate wording for the aria-label of the Section.

dgdavid commented:
> Hey, I've done some sort of simplification of the IssuesDialog component.

Thanks a lot, I think we're close to have the PR ready to merge.

> 
> However, I am not sure if it looks alright now 😓
> 
> ![image](https://private-user-images.githubusercontent.com/112288843/293465154-b7625349-5f70-41c2-867d-67aea0f8afb4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQxOTQ3NDIsIm5iZiI6MTcwNDE5NDQ0MiwicGF0aCI6Ii8xMTIyODg4NDMvMjkzNDY1MTU0LWI3NjI1MzQ5LTVmNzAtNDFjMi04NjdkLTY3YWVhMGY4YWZiNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwMlQxMTIwNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kOGRlMjQ3NzkyYWU1ZjQ0ODkyNmZmMTNiZDM0ZmRjNDY1NzFiMDgxNWY3MjhhNTIzNzk0ZGM2NzEyMWZkZGU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fWHXGI8VV_vu1dun0hIUirCK3-F2oaHodypnQvwMJLg)
> 



Actually not, It doesn't look nice now. I forget to mention previously that these warning/info icon should be removed from here too. What is more, looking at the screenshot now I overlooked the color. Not relevant when displaying these problems in the popup. Even more disturbing than helpful, IMHO. They make sense for a group of people when shown in the page, to make it more clear that there is something that requires your attention. But at this point, your attention is already on them.

> Also, unsure about some parts I've done, for example:
> 
> ```
> const titles = {
>     software: {
>       text: _("Software issues"),
>       icon: () => <Icon name="apps" />
>     },
>     product: {
>       text: _("Product issues"),
>       icon: () => <Icon name="inventory_2" />
>     },
>     storage: {
>       text: _("Storage issues"),
>       icon: () => <Icon name="hard_drive" />
>     }
>   };
> ```
> 
> Is this the right way to chose the title? Is there really a need for an icon?

No, here the icon is not needed. It's the icon of the section, but this is just a dialog with a title "Software issues" and a content, the list of issues. Feel free to get ride of the icons here.

> 
> ```
> const IssuesSections = ({ issues }) => {
>   return (
>     <Section aria-label="issues-section">
>       <IssueItems issues={issues} />
>     </Section>
>   );
> };
> ```
> 
> I've tried to simplify this part as much as possible and therefore I've removed the other prop sectionId (previously openFrom). But, I am not sure if it is maybe needed for the appropriate wording for the aria-label of the Section.

I think we can send the title as a prop and use a `_("List of issues")_` text for the Section aria-label. In any case, I'm going to comment it in the review I'm going to do right now.

Once more, thank you!

dgdavid commented on web/src/components/core/Section.jsx:
```diff
@@ -52,6 +52,8 @@ import { ValidationErrors } from "~/components/core";
  *  when user clicks on the title, used for opening a dialog.
  * @property {boolean} [loading] - Whether the section is busy loading its content or not.
+ * @property {string} [className] - Class name for section html tag.
+ * @property {string} [sectionId] - Id of the type of the section, software, product, storage, ...
```
NP: If you don't mind, we can just call it `id` here. It's ok `sectionId` in ValidationErrors and other components, but here it makes more sense just `id`. Up to you :)

Also, 

```suggestion
 * @property {string} [id] - Id of the section ("software", "product", "storage", ...)
```

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -78,39 +77,13 @@ const IssueItems = ({ issues = [] }) => {
  * @param {import ("~/client/issues").ClientsIssues} props.issues
+ * @param {string} props.sectionId - A string which indicites which issues section should be highlighted.
  */
 const IssuesSections = ({ issues }) => {
```
I'd remove this internal component, no longer needed IMHO. Just put the `<Section>` it holds directly in the place it is going to be rendered. I'd help to simplify the code even more.

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -78,39 +77,13 @@ const IssueItems = ({ issues = [] }) => {
-        }
-      />
-    </>
+    <Section aria-label="issues-section">
```
As commented before, maybe we can go for a `_("List of issues")_` or `_("List of found issues")_` instead. Something more "human readable". If you update it, do not forget about the `// TRANSLATORS:` comment to help them to understand the context of the text.

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -144,15 +118,40 @@ const IssuesContent = ({ issues }) => {
   const client = useInstallerClient();
   const { cancellablePromise } = useCancellablePromise();
-  const [notification, updateNotification] = useNotification();
+  const titles = {
```
Since we're going to drop the icon, as commented before, what about receiving the title as prop?

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -162,23 +161,29 @@ export default function IssuesPage() {
       <If
         condition={isLoading}
-        then={<SectionSkeleton numRows={4} />}
+        then={<Icon name="loading" className="icon-big" />}
```
As said, just render the `<Section>` here. `IssuesContent` internal component it's no longer needed. And maybe place the loading inside the section, 

```jsx

<Section aria-label="...">
  <If condition={isLoading} then={...} />
  <IssueItems issues={issues} />
</Section>
```

Or a better approach if you found it while coding :)

dgdavid commented on web/src/components/core/ValidationErrors.jsx:
```diff
@@ -77,27 +60,22 @@ const ValidationErrors = ({ title = _("Errors"), errors }) => {
           <button
```
Not related with your changes, but do you mind removing the `else` above and just returning the content? :innocent:  Less indentation and easy to read by just relaying on the early return when errors.length === 1.

dgdavid reviewed (commented):
@balsa-asanovic, 

We're close to the merge :) Keep going! :muscle: :man_technologist: :keyboard:

dgdavid commented on web/src/components/core/ValidationErrors.jsx:
```diff
@@ -22,45 +22,28 @@
-const ValidationErrors = ({ title = _("Errors"), errors }) => {
-  const [popoverVisible, setPopoverVisible] = useState(false);
+const ValidationErrors = ({ errors, sectionId }) => {
+  const [showIssuesPopUp, setshowIssuesPopUp] = useState(false);
```
NP: better to use camelCase all the time, 

```suggestion
  const [showIssuesPopUp, setShowIssuesPopUp] = useState(false);
```

Spotted by the spell checker, https://github.com/openSUSE/agama/actions/runs/7364595390/job/20090333162?pr=886#step:6:12

Which you can run locally with `npm run cspell` if you want.

dgdavid commented on web/src/components/core/IssuesDialog.jsx:
```diff
@@ -78,39 +77,13 @@ const IssueItems = ({ issues = [] }) => {
  *
  * @param {object} props
  * @param {import ("~/client/issues").ClientsIssues} props.issues
+ * @param {string} props.sectionId - A string which indicites which issues section should be highlighted.
```
Also spot by cspell, 

```suggestion
 * @param {string} props.sectionId - A string which indicates which issues section should be highlighted.
```

dgdavid commented:
Also, is pending to drop _dead_ code:

* src/components/core/IssuesLink.jsx
* src/components/core/IssuesLink.test.jsx
* src/context/notification.js, which implies updating few files for stop using `NotificationProvider`. Namely, 
  - src/components/core/IssuesDialog.test.jsx (which I guess will change a lot once it gets updated for reflecting changes done in the PR)
  - src/context/agama.jsx
  - src/test-utils.js

I discussed it last week with @imobachgs. We prefer dropping dead code and restoring it back if needed in the future instead of keeping it there "just in case".

balsa-asanovic commented:
Hey @dgdavid,

I think I covered all the comments.

The test file for the IssuesDialog will require some more work 😄, I've put just two basic tests for now.

dgdavid reviewed (commented):
@balsa-asanovic 

Thanks a lot for addressing the latest set of comments/suggestions. Much appreciated.

**I hope you don't mind I've sent some commits this morning**. Otherwise, please excuse me for doing so. Reason was not other than speeding up this PR in order to allow you jump to another one if you want :innocent: I thought it was quickly than writing another bunch of comments. Furthermore, the latest ones are things that I realized when working in the formers.

I'll explain some things in a while, in a comment below.

BTW, time of adding the entry in the changelog :)

dgdavid commented:
> BTW, time of adding the entry in the changelog :)

And also, if you don't mind, time for updating the PR description since it is a bit outdated.

 Not sure if you noticed it, but we configured the repo for including it as the commit description by default (See https://github.blog/changelog/2022-08-23-new-options-for-controlling-the-default-commit-message-when-merging-a-pull-request/ to know more)

dgdavid commented:
> I'll explain some things in a while, in a comment below.

I've made two changes of note, 

* Drop the core/`Section` component from the `IssuesDialog`

  It was me who proposed using it, my fault. However, when taking a look at the final result, I realized that visually it does not look quite well when rendered inside a `Popup` without using a section title and icon because at this time `Section` reserves the space for the icon. Not sure if in the future the component will be more flexible and allows conditionally to reserve the spacer or not, but for now not using it here is the straight forward solution. 


  | Before | After |
  |-|-|
  | ![Screen Shot 2024-01-04 at 10 12 00](https://github.com/openSUSE/agama/assets/1691872/98601d0a-7c19-48af-b9da-428bc4a04f59) | ![Screen Shot 2024-01-04 at 10 50 39](https://github.com/openSUSE/agama/assets/1691872/e2b43bbc-65cb-42bb-aa64-092eb8e2b0f5) |

  Note that, additionally, now issues are wrapped by a plain HTML `ul` element instead of a collection of nested `div`s. IMHO, semantically more accurate: it is a list of issues, at the end.


* Make the `ValidationErrors` component able to infer the _issues area_ from the given _sectionId_.

  This is needed because the `IssuesDialog` can be now open from sections _sub-headers_ and it will display the errors given _area_ based on the id of the section. Maybe is something to improve, at least semantically. But in another PR. For now, the change introduced at https://github.com/openSUSE/agama/pull/886/commits/e1b7f79e75a70cf4612caeb53d4c097a1e328091 allows displaying the issues for the same _area_ from different sections without forcing all of them to have the same id, which will be wrong if sections are rendered in the same page. For example, from the `StorePage` the `Actions` section allows now opening such a dialog by using a _meaningful_ `storage-actions` id. A bit tricky, indeed, but enough for considering this PR ready.

  | Before (without id) | After (using `storage-actions`|
  |-|-|
  | ![Screen Shot 2024-01-04 at 11 08 07](https://github.com/openSUSE/agama/assets/1691872/26456c21-4d6d-49d4-9e71-756aa051a0c2) | ![Screen Shot 2024-01-04 at 11 08 13](https://github.com/openSUSE/agama/assets/1691872/c7036d89-3413-4063-8519-e5fd05f12509) |

balsa-asanovic commented:
> I hope you don't mind I've sent some commits this morning. Otherwise, please excuse me for doing so.

Of course, I don't mind at all.

> I realized that visually it does not look quite well when rendered inside a Popup

This was exactly something I noticed last night when making the changes.

I thought about suggesting using some sort of bullets, but at the end didn't leave a comment as to not extend the PR more.

Glad to see you went in the similar direction. 👍

dgdavid commented:
> This was exactly something I noticed last night when making the changes.
> 
> I thought about suggesting using some sort of bullets, but at the end didn't leave a comment as to not extend the PR more.

Please, in the future, do not hesitate on making as many as comment or suggestion you want. I'm pretty sure they will help the project to go in the right direction. Don't worry about the "pr extension", we can help to move it forward at any time.

> 
> Glad to see you went in the similar direction. 👍

Glad to know you had realize as well.

balsa-asanovic commented:
Okay, so I removed the Issues Dialog route from index file.

I've updated the PR title and description, feel free to adjust this further if you notice something wrong or missing.

Also, I've added the entry to the changes file.

dgdavid reviewed (approved):
@balsa-asanovic,

Thanks for your hard work on this :100: 

Definitely, it has been a PR quite longer than expected, mostly because we changed our minds about the final result while reviewing your remarkable contribution. I understand how frustrating that can be, but you've been incredibly patient and broad-minded understanding we are working on live project where some things are re-defined _on the fly_. Your willingness to change almost everything again surprised me. I hope at least you've had good coding moments and learned something along the way.

As you have written, there is still a bit of work to do here and there. Nevertheless, the **PR looks good to me** :+1:  as it is, especially taking into account that these tweaks will come once we have them more clear by gathering feedback from both developers and users.

Needless to say, looking forward to reviewing your next PR for the Agama project. I'm curious about which issues you're gonna pick :stuck_out_tongue_winking_eye: 

See ya.

balsa-asanovic commented:
Hey @dgdavid,

Thanks a lot for the kind words.

>  I understand how frustrating that can be

Not at all, as you said Agama is a live project and changing things on the fly is quite alright.

> I hope at least you've had good coding moments and learned something along the way.

I enjoyed working on this and picked up quite a few new things (I always do here at Agama).

In large part this is due to your reviews and very detailed comments, I realize that sometimes just doing something yourself would be easier and faster, so thank you a lot for you willingness to always help.

> Needless to say, looking forward to reviewing your next PR for the Agama project. I'm curious about which issues you're gonna pick 😜

Haha, if you have some good idea about an issue (or a planned feature) on which I could work, please do make a suggestions.

If not, here's something of interest I noticed after a quick read through the issues:

#870 This seems interesting, however I'd need a bit more of an explanation than currently available in the issue description.

#112 This is an older issue. I mentioned already at one point that I could take it, it seems that it is still available.

Some other recent issues seem to be already taken by new contributors.

But, again, if you have a better idea than this, please do say, these are just suggestions.

balsa-asanovic commented:
Hey @dgdavid,

If you find the time would you mind checking this out:

> Haha, if you have some good idea about an issue (or a planned feature) on which I could work, please do make a suggestions.
> 
> If not, here's something of interest I noticed after a quick read through the issues:
> 
> https://github.com/openSUSE/agama/issues/870 This seems interesting, however I'd need a bit more of an explanation than currently available in the issue description.
> 
> https://github.com/openSUSE/agama/issues/112 This is an older issue. I mentioned already at one point that I could take it, it seems that it is still available.
> 
> Some other recent issues seem to be already taken by new contributors.
> 
> But, again, if you have a better idea than this, please do say, these are just suggestions.

Thanks

dgdavid commented:
Sorry @balsa-asanovic, I had forgotten to answer to your question. Thanks for the friendly ping :)

After discussing it with @imobachgs, we believe the #112 is the good candidate since, 

* It's old enough :see_no_evil: 
* It's completely about the front-end
* You already have provided a lot of feedback there :100: 
* As you pointed out, #870 was not well-defined. In fact, I've moved it to a discussion 

Of course, ask whatever doubt you have.

ancorgs commented:
JFYI, this pull request was mentioned in the [announcement of Agama 8](https://yast.opensuse.org/blog/2024-05-17/agama-8).

Thanks for your contribution!
